### PR TITLE
Add proof for ECDSA.

### PR DESCRIPTION
**Pull request by andreistefanescu**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: SAW/proof/ECDSA/evp-function-specs.saw [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+let EVP_DigestInit_spec op = do {
+  ctx_ptr <- crucible_alloc (llvm_struct "struct.env_md_ctx_st");
+  points_to_env_md_ctx_st ctx_ptr crucible_null crucible_null;
+
+  type_ptr <- crucible_alloc_readonly (llvm_struct "struct.env_md_st");
+  points_to_env_md_st type_ptr;
+
+  pkey_ptr <- crucible_alloc (llvm_struct "struct.evp_pkey_st");
+  //(ec_key_ptr, _pub_key, _priv_key) <- pointer_to_fresh_ec_key_st op;
```
Comment (andreistefanescu) [inline SAW/proof/ECDSA/evp-function-specs.saw:27]: `EVP_DigestInit` works even if `ec_key` is uninitialized or `NULL`. Should the spec require that `ec_key` is initialized?

File: SAW/proof/ECDSA/evp-function-specs.saw [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  sha512_state_ptr <- crucible_alloc (llvm_struct "struct.sha512_state_st");
+  sha512_state <- fresh_sha512_state_st "sha512_state" num;
+  points_to_sha512_state_st sha512_state_ptr sha512_state num;
+
+  evp_pkey_ctx_ptr <- crucible_alloc_readonly (llvm_struct "struct.evp_pkey_ctx_st");
+  //pkey_ptr <- crucible_alloc_readonly (llvm_struct "struct.evp_pkey_st");
+  //(ec_key_ptr, _pub_key, _priv_key) <- pointer_to_fresh_ec_key_st op;
+  //points_to_evp_pkey_st pkey_ptr ec_key_ptr;
+  //ec_pkey_ctx_ptr <- crucible_alloc_readonly (llvm_struct "struct.EC_PKEY_CTX");
+  //points_to_EC_PKEY_CTX ec_pkey_ctx_ptr digest_ptr;
```
Comment (andreistefanescu) [inline SAW/proof/ECDSA/evp-function-specs.saw:61]: `EVP_DigestUpdate` works even if `pkey` and `pkey_ctx` are uninitialized or `NULL`. Should the spec require that they are initialized?

File: SAW/proof/ECDSA/ECDSA.saw [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+  (do {
+    unfolding ["ECDSA_sign"];
+    unfolding ["jacobianToMontBV", "jacobianFromMontBV", "jacobianToMont", "jacobianFromMont", "jacobianToBV", "jacobianFromBV"];
+    simplify (cryptol_ss ());
+    simplify (addsimp scalarBV_thm empty_ss);
+    simplify (addsimp fieldElementBV_thm empty_ss);
+    simplify (addsimps mont_thms empty_ss);
+    simplify (addsimp jacobian_affine_0_thm empty_ss);
+    simplify (addsimps mod_thms empty_ss);
+    print_goal;
```
Comment (apetcher-amazon) [inline SAW/proof/ECDSA/ECDSA.saw:199]: Can you remove the print_goal statements? There are others in this file, so make sure to get all of them.

**Review by apetcher-amazon — COMMENTED** [review]
Please update README.md and SAW/README.md to describe the ECDSA proof.

File: SAW/proof/EC/EC.saw [hunk 4]
```diff
@@ -0,0 +1,632 @@
...
+
+
+////////////////////////////////////////////////////////////////////////////////
+// Proof commands
+
+EC_GFp_mont_method_ov <- crucible_llvm_unsafe_assume_spec
+  m
+  "EC_GFp_mont_method"
+  EC_GFp_mont_method_spec;
+EC_GFp_nistz256_method_ov <- crucible_llvm_unsafe_assume_spec
```
[the remark this answers is not in the document]
Comment (andreistefanescu) [reply SAW/proof/EC/EC.saw:415]: the code that initializes the builtin curves calls these functions in order to get the method tables for the curves. These functions simply return the globals that store the method tables.

File: SAW/README.md [hunk 5]
```diff
@@ -72,3 +73,17 @@ The AES_(un)wrap_key_* functions are verified to have the following properties r
...
+
+### ECDSA
+
+The EVP_DigestSign*/EVP_DigestVerify* functions are verified to have the following properties related to ECDSA using P-384 and SHA-384. For more detailed specifications, see [evp-function-specs.saw](proof/ECDSA/evp-function-specs.saw). BLOCK_LENGTH is the block length of the hash function, in bytes. MAX_SIGNATURE_LENGTH is the digest length of the hash function, in bytes. For ECDSA with P-384 and SHA-384, BLOCK_LENGTH is 64 and MAX_SIGNATURE_LENGTH is 104.
```
Comment (apetcher-amazon) [inline SAW/README.md:79]: Looks like the description of MAX_SIGNATURE_LENGTH is incorrect.

File: SAW/proof/ECDSA/memory.saw [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+*/
+
```
[the remark this answers is not in the document]
Comment (apetcher-amazon) [reply SAW/proof/ECDSA/memory.saw:5]: I need more information to understand where the numbers came from in each override. For example, in "(OPENSSL_realloc_spec 0x34 0x68)", why 0x34 and 0x68? Are these values fixed, or do they depend on the input length, hash length, or curve? Mostly, I want to be able to easily maintain this file when we use different settings, so use constants (ideally calculated from other constants), if possible.

**Review by apetcher-amazon — APPROVED** [review]

**Pull request merged**